use-cases/xtclim/src/trainer.py

```diff
@@ -14,56 +14,114 @@
+)
+from itwinai.loggers import Logger
+from itwinai.torch.config import TrainingConfiguration
 
-from itwinai.components import Trainer, monitor_exec
 import model
 from torch.utils.data import DataLoader
 from torchvision.utils import make_grid
 from engine import train, validate
 from utils import save_reconstructed_images, save_loss_plot, save_ex
-from initialization import device, beta, criterion
+from initialization import beta, criterion, pixel_wise_criterion
 
-class TorchTrainer(Trainer):
+class XTClimTrainer(TorchTrainer):
     def __init__(
             self,
             epochs: int,
             batch_size: int,
-            lr: float
+            lr: float,
+            strategy: Literal["ddp", "deepspeed", "horovod"] = 'ddp',
+            save_best: bool = True,
+            logger: Optional[Logger] = None
     ):
-        super().__init__()
-        self.epochs = epochs
+        super().__init__(
+            epochs=epochs,
+            config={},
+            strategy=strategy,
+            logger=logger
+        )
         self.batch_size = batch_size
+        self.epochs = epochs
         self.lr = lr
+        # Global training configuration
+        self.config = TrainingConfiguration(
+            batch_size=self.batch_size,
+            save_best=save_best,
+        )
+
+    def final_loss(self, bce_loss, mu, logvar, beta=0.1):
+        """
+        Adds up reconstruction loss (BCELoss) and Kullback-Leibler divergence.
+        KL-Divergence = 0.5 * sum(1 + log(sigma^2) - mu^2 - sigma^2)
+
+        Parameters:
+        bce_loss: recontruction loss
+        mu: the mean from the latent vector
+        logvar: log variance from the latent vector
+        beta: weight over the KL-Divergence
+        """
+        BCE = bce_loss
+        KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
+        return BCE + beta*KLD
+
 
     @monitor_exec
     def execute(self):
+
         # pick the season to study among:
         # '' (none, i.e. full dataset), 'winter_', 'spring_', 'summer_', 'autumn_'
         seasons = ["winter_", "spring_", "summer_", "autumn_"]
 
         # number of members used for the training of the network
         n_memb = 1
 
-        # initialize learning parameters
-        #lr0 = 0.001
-        #batch_size = 64
-        #epochs = 100
         #early stopping parameters
         stop_delta = 0.01  # under 1% improvement consider the model starts converging
         patience = 15  # wait for a few epochs to be sure before actually stopping
         early_count = 0  # count when validation loss < stop_delta
         old_valid_loss = 0  # keep track of validation loss at t-1
 
+        # Initialize distributed backend
+        self._init_distributed_strategy()
         for season in seasons:
-
             # initialize the model
-            cvae_model = model.ConvVAE().to(device)
+            cvae_model = model.ConvVAE()
             optimizer = optim.Adam(cvae_model.parameters(), lr=self.lr)
```

Comment:
lr should be retrieved from `self.config`

Reply:
Done